justfile

```diff
@@ -29,10 +29,11 @@ gen-bridge: build
                 -d {{frb_flutter}}/lib/bridge_generated.dart \
                 --dart-decl-output {{frb_flutter}}/lib/bridge_definitions.dart \
                 -c {{frb_flutter}}/ios/Runner/bridge_generated.h \
-                -c {{frb_flutter}}/macos/Runner/bridge_generated.h \
+                # -c {{frb_flutter}}/macos/Runner/bridge_generated.h \
```

Comment:
Ah this is what I mean. We should really generate to both ios and macos!

Reply:
If it is necessary, why CI tests for IOs and MacOs are passed?

Reply:
> Ah this is what I mean. We should really generate to both ios and macos! So this line cannot be commented - otherwise, the macos header file will be stale soon, and cause problems

If it is necessary, why CI tests for IOs and MacOs are passed?

Reply:
because there are *stale* bridge_generated.h on that location :/

Reply:
First, I still have no idea what "stale" means here. What kind of problem would occur, would this result in a failed compilation?

Second, I read [documentation ](https://cjycode.com/flutter_rust_bridge/integrate/ios_headers.html)here, it seems that the c header file has to be put in path `ios/Runner/` for building IOS APP or `macos/Runner/` for MacOS APP, and users have to specify the `c-output` manually, right (please correct me if I am wrong)?  

If this is the case, my suggestion is to add a new flag `apple` while following the logic of this pr. Then, once the flag `apple` is provided:

1. if no `c-output` flag is provided, frb could generate c header(s) in `ios/Runner/` and `macos/Runner/` respectively with the same name in each APPLE path, which corresponds to the same rust block--- Say, there are 2 rust blocks, then `c_output_01.h` and `c_output_02.h` would be provided in both `ios/Runner/` and `macos/Runner/`(altogether 4 files).
2. if the `c-output` flag is provided, then like for other target platforms(Android, Windows),  generation would still fail if the provided number of fields in  `c-output`  is not consistent with that of `rust-input`. Otherwise, after generating the c file of path/name specified in `c-ouput`, frb could then generate (ALL) c file(S) in the above stated 2 APPLE paths with the specified name(s).

The logic here is:
no matter what platform, each field of  `c-output` should always have the same meaning. 
For special platforms (IOS/MacOS), just tell frb with a new flag, then frb would handle all other dirty work.

Reply:
> 

just 1 thing to clarify, to build IOS or macos APP, the coding platform has to be MacOS, right?  If this is the case, then the new flag is no needed, rust could detect running OS by :
```rust
use std::env;

fn main() {
    let os_string = env::consts().os;
    println!("Current OS String: {}", os_string);
}
```

Reply:
> First, I still have no idea what "stale" means here. What kind of problem would occur, would this result in a failed compilation?

Suppose user changes api.rs or something like that, and trigger frb codegen. Then, if you comment out the macos one, we will see ios/bridge.h updated, **but macos/bridge.h is still old and not updated at all**. Then, when you build macos app, you will be using the new api.rs **and the old bridge.h** so it can cause trouble

Reply:
> If this is the case, my suggestion is to add a new flag apple while following the logic of this pr.

Well IMHO it may not be perfect. Because frb should not be bound to Flutter and platforms too tightly. It should work well with even pure dart, and whatever folder structure that people like. It may even generate bridge.h into android/windows/whatever platform as long as the user needs it.

Reply:
> If it is necessary, why CI tests for IOS and MacOS are passed?

Since we only pass for non-multiple APIs before, when there was only one API, we needed to send the header to ios and mac separately
Of course, you can give it to Linux if you want.

Just my understanding.
z_z

Reply:
> It may even generate bridge.h into android/windows/whatever platform as long as the user needs it.

With the design of this pr,  it is feasible for users to generate the c header file into any custom path by flag `c-output`. The issue is just about MacOS and IOS, for which the c head file is essential for building app with flutter/dart, if I am not mistaken.

> Because frb should not be bound to Flutter and platforms too tightly. It should work well with even pure dart

Here some things I need to clarify:
1. For platforms like Windows, web, and Android, the c head file during generating is not needed after generating the corresponding dart file? 
2. for building a flutter app targeting IOS/MacOS, the c head file is essential, right? If so, is the path `ios/Runner/` or `macos/Runner/` the compulsory path?
3. for building an app with pure dart? Does the c head file necessary? If so, where the file should be put?

Reply:
> For platforms like Windows, web, and Android , is the c head file during generating still needed after generating the corresponding dart file? (In my test, I think the answer is no)
> for building an app with pure dart? Does the c head file necessary? If so, where the file should be put?

Well my point is that, even if *currently our scenario* does only require macos/ios and not other platforms to have .h, we may not rely on it. If we rely on it, we are essentially writing the frb codebase in a way such that *frb mandatory knows what is Flutter platforms*, while I have claimed in README that this package is pure-dart compatible.

In other words, "high cohesion, low coupling" - IMHO we should not couple our frb with flutter platform knowledge

Reply:
> > For platforms like Windows, web, and Android , is the c head file during generating still needed after generating the corresponding dart file? (In my test, I think the answer is no)
> > for building an app with pure dart? Does the c head file necessary? If so, where the file should be put?
> 
> Well my point is that, even if _currently our scenario_ does only require macos/ios and not other platforms to have .h, we may not rely on it. If we rely on it, we are essentially writing the frb codebase in a way such that _frb mandatory knows what is Flutter platforms_, while I have claimed in README that this package is pure-dart compatible.
> 
> In other words, "high cohesion, low coupling" - IMHO we should not couple our frb with flutter platform knowledge

So what is your suggestion to solve these issues?

Reply:
For example (may be good or bad), for 2 api.rs and 2 platforms (ios and macos), let users specify 2x2=4 locations

Reply:
> For example (may be good or bad), for 2 api.rs and 2 platforms (ios and macos), let users specify 2x2=4 locations

what about this:
1. take flag `c-ouput` follows the logic of this pr--each filed represent one API block;
2. take a new flag, say  `extra-c-output-path`, for bak all c files into specific path;

Example 1:
there are 2 API blocks, and users want to put c files in `pathA`, `pathB` and `pathC`,
then command could be:
```
...
-c-output "pathA/api1.h" "pathA/api2.h"
-extra-c-output-path  "pathB"  "pathC"
...
```
Here in `pathA`, `pathB`, and `pathC`, `api1.h` and `api2.h` are contained relatively.

Example 2:
there are 2 API blocks, the name of c file is no cared, but users want to put them in  `pathA`,
then command could be:
```
...
# no -c-output
-extra-c-output-path  "pathA"
...
```
In this case, all generated files would bak into `pathA`, though the names are  arbitrary.

Reply:
LGTM

Reply:
> LGTM :)

I've commit accordingly in ba004911c64c850118c4ca5fcc914761cae51a08, but it is weird that the CI test is not triggerred.